_upgrading/sopel-8.md

```diff
@@ -0,0 +1,420 @@
+
+The following parts of Sopel's API are now expressed as some type of `Enum`:
+
+* `sopel.formatting.colors`
+* `sopel.tools.events`
+
+### Identifier casemapping
+
+Any reasonably modern IRC server advertises its method for normalizing nicknames
+to every client that connects. Sopel 8 builds on the `bot.isupport` mechanism
+added in Sopel 7 and makes use of the advertised `CASEMAPPING` value when
+comparing and normalizing nicknames itself.
+
+To support handling this under the hood, the bot's `config.core.nick` setting is
+now stored and returned as a normal `str`. It and any other strings that you'd
+like to treat as nicknames should be passed through the `bot.make_identifier()`
+method to get the same `Identifier` type you're used to from Sopel 7, configured
+to use the IRC server's advertised casemapping method.
+
+### Time-handling changes
+
+`trigger.time` is now an "aware" `datetime` object, meaning it has a UTC offset
+associated with it. Comparison or arithmetic operations between "aware" and
+"naive" `datetime` objects are not allowed; code that manipulates `trigger.time`
+values will need to be updated for Sopel 8.
+
+The fallback format string used by `sopel.tools.time.format_time()` if no other
+source provides one changed from outputting the timezone name (`UTC`) to the UTC
+offset (`+0000`).
+
+`format_time()`'s handling of "aware" and "naive" `datetime`s was also improved,
+but those changes should be transparent to both users and plugin developers.
+
+`sopel.tools.time.validate_timezone()` now also raises a `ValueError` even if
+the timezone to be validated is `None`. This was previously a special case that
+_returned_ `None`.
+
+### Database changes
+
+`bot.db.get_nick_id()` no longer creates a new ID by default if its `create`
+parameter is unspecified.
+
+### IRC event changes
+
+The event names `RPL_INVITELIST` and `RPL_ENDOFINVITELIST` have been updated per
+clarifications from the living-specification project at modern.ircdocs.horse.
+`RPL_INVEXLIST` and `RPL_ENDOFINVEXLIST` were added to the list of named events
+that Sopel knows about.
+
+Plugin code might need to be updated as a result of these changes, but keep in
+mind that [different IRC servers mean different things][invitelist-madness] when
+they send these events. If in doubt, specify a raw numeric value (e.g. `'346'`).
+Writing truly cross-network plugins that react to these events should be
+undertaken _very_ carefully.
+
+  [invitelist-madness]: https://github.com/ircdocs/modern-irc/issues/42
+
+### IRC connection status monitoring
+
+Previously a simple `bool` value, `bot.connection_registered` is now a property
+combining status information from across the bot's subsystems to answer the very
+important question, "Is the `bot` actually registered to an IRC network?" That
+is, Sopel not only has a _connection_ open, but the IRC server has _accepted_
+the bot as a _client_ and it's possible to _send commands_ to the IRC server.
```

Comment:
Sharing here a reword I did of this in my initial review some months ago. I don't feel as strongly now as I did about it then, though.

```suggestion
`bot.connection_registered` now indicates that a _connection_ is open, the 
IRC server has _accepted_ the bot as a client, and commands can be sent.
Previously, this boolean value indicated only that a connection was open. This
attribute is now read-only and cannot be modified.
```

Reply:
Going to take my own stab at this one, using yours as a starting point:

```suggestion
`bot.connection_registered` now indicates that a _connection_ is open, the 
IRC server has _accepted_ the bot as a client, and the bot can _send commands_.
Previously, this boolean value indicated only that a connection was open.
```

Removed the disclaimer that it's read-only _now_ because no one outside core should have been writing to it anyway. We've just codified the implicit contract by making it into a `@property`.